proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+//       float4 g_vSomeConstantB;
+// };
+
+struct TestPushConstant_t
+{
+      uint64_t m_nBufferDeviceAddress;  // GlobalsTest_t
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+    float4 vTest = vk::RawBufferLoad<float4>(g_PushConstants.m_nBufferDeviceAddress + 16);
+      
+      return vTest;
+}
+```
+
+The SPIR-V for this shader can be seen in Appendix A. Note the lack of logical context for the accessed buffer i.e. no declaration for the underlying structure GlobalsTest_t as is generated for other buffers.
+
+There is another way to use RawBufferLoad which does allow logical selection of the buffer fields, but it inefficiently loads the entire buffer to do it. See https://github.com/microsoft/DirectXShaderCompiler/issues/4986.
+
+The goal of this proposal is to have a solution that meets the following requirements:
+
+* Removes the need for having to manually or automatically generate offsets to load structured data with BufferDeviceAddress.
+* Get equivalent tooling functionality as is provided by the buffer reference feature in GLSL.  Namely, tools like RenderDoc are able to introspect the type information such that its buffer inspection and shader debugger are able to properly understand and represent the type of the data.
+* Make it possible through SPIR-V reflection to determine which members of a struct accessed by BufferDeviceAddress are statically referenced and at what offset.  This is already possible for other data like cbuffers in order for shader tooling to be able to identify which elements are used and where to put them.
+
+## Proposed solution
+
+Our solution is to add a new builtin type in the vk namespace that is a pointer to a buffer of a given type, `vk::buffer_ptr<T,A>`. The template argument `T` must be a struct. `A` must be an integer and is the alignment in bytes of the pointer. If `A` is not specified, the alignment is assumed to be 16 bytes.
+
+This new type will have the following operations
+
+* Copy assignment and copy construction - These copy the value of the pointer from one variable to another.
+* Accessing members of T - This is done using the `.` operator.
+* Conversion to uint64_t - Return the integer representation of the address it points to. This can be used for null pointer checks.
+
+Note the operations that are not allowed:
+
+* There is no default construction. Every vk::buffer_ptr<T> is either contained in a global resource (like a cbuffer, ubo, or ssbo), or it must be constructed using the copy constructor. 
+* There is no equivalent to the `*` operator that will dereference the entire struct.
+* There is no conversion from uint64_t to vk:buffer_ptr.
+* There is no explicit pointer arithmetic. All addressing is implicitly done using the `.` pointer, or indexing an array in the struct T.
+* The comparison operators == and != are not supported for buffer pointers. Comparison to null pointer should be done by first converting to uint64_t.
```

Comment:
Should we convert to `uint64_t` for comparison to null or should we support a boolean conversion for validity?

Reply:
I’m not sure how I feel about introducing a `nullptr` construct… I need to think on that.

Reply:
I personally like the boolean conversion idea or we could add a member function `IsNull()`. I find the member function more readable.

Reply:
The only reason for the convert to uint64_t is for the null check, so replacing the convert with an IsNull() method would be acceptable. It would also remove the dependency on the availability of uint64_t.

Reply:
Yea, I think `IsNull()` is much cleaner than supporting conversion to `uint64_t`. We really don't want/need to expose the address value directly.

Reply:
Yes, fine with `IsNull()`. Just curious about any future divergence: cuda/opencl have pointers and `!= NULL`, so why not other shaders in a few years?  Using a bool operator with `!myBufferPointer` at least allows us be "compatible" with a more native pointer syntax, if we need one in the future.

Reply:
I don't think we've ever made convergence with cuda/opencl a goal for HLSL. That said, you're probably right that we will want to support pointers in the future in HLSL.

The problem with a `bool` conversion operator is actually an existing issue in HLSL. Overloaded operators for user defined types cannot be used for Implicit conversions (there's a [bug](https://github.com/microsoft/DirectXShaderCompiler/issues/4096) tracking this). I posted a fly-by comment on that bug back in February because I think I spotted the problem, but I'm unsure if that can be safely changed in HLSL 2021.

If it can't be fixed safely in HLSL 2021, then a conversion operator would also need to be part of the HLSL 202x feature set where I have more extensive fixes for overload resolution planned.